src/blob_gc_job.cc

```diff
@@ -582,69 +582,80 @@ bool BlobGCJob::IsShutingDown() {
-    for (; record_iter->Valid(); record_iter->Next()) {
-      if (IsShutingDown()) {
-        s = Status::ShutdownInProgress();
-        return s;
+Status BlobGCJob::DigHoleFile(BlobFileMeta *input) {
+  Status s;
+  //open file
+  std::unique_ptr<PosixRandomRWFile> file;// TODO(@DorianZheng) set read ahead size
+  s = OpenBlobFile(input->file_number(), 0, db_options_, env_options_, env_, &file);
+  if (!s.ok()) {
+    return s;
+  }
+  auto record_iter = new BlobFileIterator(
+      std::move(file), input->file_number(), input->file_size(),
+      blob_gc_->titan_cf_options());
+  //pre size
+  uint64_t before_size = 0, after_size = 0;
+  record_iter->GetFileRealSize(&before_size);
+  //for each block-clusters
+  for (record_iter->SeekToFirst(); record_iter->Valid();) {
+    if (IsShutingDown()) {
+      s = Status::ShutdownInProgress();
+      return s;
+    }
+    //read first record in block-cluster and get hole size
+    BlobIndex blob_index = record_iter->GetBlobIndex();
+    BlobHandle blob_handle = blob_index.blob_handle;
+    uint64_t record_end = blob_handle.offset + blob_handle.size;
+    const uint64_t hole_start = blob_handle.offset;
+    const uint64_t hole_end = (record_end + block_size_) / block_size_ * block_size_;
+    bool enableHole = true;
+
+    while (record_end <= hole_end) {
+      //judge
+      if (enableHole) {
+        bool discardable = false;
+        s = DiscardEntry(record_iter->key(), blob_index, &discardable);
+        if (!s.ok()) {
+          break;
+        }
+        if (!discardable) {//There is valid record.
+          enableHole = false;
+        }
       }
-
-      BlobIndex blob_index = record_iter->GetBlobIndex();
-      bool discardable = false;
-      s = DiscardEntry(record_iter->key(), blob_index, &discardable);
-      if (!s.ok()) {
+      //read next record
+      record_iter->Next();
+      if (!record_iter->Valid()) {
         break;
-      }
-
-      const auto& blob_handle = record_iter->GetBlobIndex().blob_handle;
-      int64_t blob_tail = blob_handle.offset + blob_handle.size;
-      if (discardable) {  // cur key is discardable so skip
-        last_discadable_tail = blob_tail;
       } else {
-        cur_valid_head = blob_handle.offset;
-        if (cur_valid_head != last_valid_tail) {
-          record_iter->PunchHole(last_valid_tail,
-                                 cur_valid_head - last_valid_tail);
-        }
-        last_valid_tail = blob_tail;
+        blob_index = record_iter->GetBlobIndex();
+        blob_handle = record_iter->GetBlobIndex().blob_handle;
+        record_end = blob_handle.offset + blob_handle.size;
       }
     }
 
-    if (!record_iter->status().ok()) {
-      return record_iter->status();
-    }
-    if (last_valid_tail < last_discadable_tail) {
-      record_iter->PunchHole(last_valid_tail,
-                             last_discadable_tail - last_valid_tail);
+    if (enableHole && record_iter->status().ok() && s.ok()) {
```

Comment:
In last block, record_iter maybe invalid, can we still punch the hole?

Reply:
We discussed the problem, when record_iter access the last block, it is invalid and status is ok, so it can punch the hole..